ocaml-lsp-server/src/ocamlformat_rpc_lib.ml

```diff
@@ -0,0 +1,220 @@
+      | Ok (List [ Atom "Config"; List l ]) ->
+        let c =
+          List.fold_left
+            ~f:(fun acc -> function
+              | List [ Atom name; Atom value ] -> (name, value) :: acc
+              | _ -> acc)
+            ~init:[] l
+          |> List.rev
+        in
+        `Config c
+      | Ok (List [ Atom "Error"; Atom x ]) -> `Error x
+      | Ok (Atom "Halt") -> `Halt
+      | Ok _ -> `Unknown
+      | Error _msg -> `Halt
+
+    let to_sexp =
+      let open Sexp in
+      function
+      | `Format x -> List [ Atom "Format"; Atom x ]
+      | `Config c ->
+        let l =
+          List.map ~f:(fun (name, value) -> List [ Atom name; Atom value ]) c
+        in
+        List [ Atom "Config"; List l ]
+      | `Error x -> List [ Atom "Error"; Atom x ]
+      | `Halt -> Atom "Halt"
+      | _ -> assert false
+
+    let output channel t =
+      to_sexp t |> Csexp.to_channel channel;
+      Stdlib.flush channel
+  end
+
+  module Client = struct
+    type t =
+      { pid : int
+      ; input : in_channel
+      ; output : out_channel
+      }
+
+    type cmd = Command.t
+
+    let pid t = t.pid
+
+    let mk ~pid input output = { pid; input; output }
+
+    let query command t =
+      Command.output t.output command;
+      Command.read_input t.input
+
+    let halt t =
+      match
+        Command.output t.output `Halt;
+        close_in t.input;
+        close_out t.output
+      with
+      | exception _ -> Error (`Msg "failing to close connection to server")
+      | () -> Ok ()
+
+    let config c t =
+      match query (`Config c) t with
+      | `Config _ -> Ok ()
+      | `Error msg -> Error (`Msg msg)
+      | _ -> Error (`Msg "failing to set configuration: unknown error")
+
+    let format x t =
+      match query (`Format x) t with
+      | `Format x -> Ok x
+      | `Error msg -> Error (`Msg msg)
+      | _ -> Error (`Msg "failing to format input: unknown error")
+  end
+end
+
+type client = [ `V1 of V1.Client.t ]
+
+let get_client ~pid input output = function
+  | "v1"
+  | "V1" ->
+    Some (`V1 (V1.Client.mk ~pid input output))
+  | _ -> None
+
+let get_client_exn ~pid input output x =
+  match get_client ~pid input output x with
+  | Some x -> Ok x
+  | None -> failwith "impossible"
+
+let pick_client ~pid input output versions =
+  let rec aux = function
+    | [] -> Error (`Msg "Version negociation failed")
+    | latest :: others -> (
+      let version = `Version latest in
+      Csexp.to_channel output (Init.to_sexp version);
+      flush output;
+      match Init.read_input input with
+      | `Version v when v = latest -> get_client_exn ~pid input output v
+      | `Version v -> (
+        match others with
+        | h :: _ when v = h -> get_client_exn ~pid input output v
+        | _ -> aux others )
+      | `Unknown -> aux others
+      | `Halt ->
+        Error
+          (`Msg
+            "OCamlformat-RPC did not respond. Check that a compatible version \
```

Comment:
We should probably mentioned what minimum version is required.

Reply:
cc @gpetiot

Reply:
done